File: src/Java/algorithms/treeandgraph/FlattenBinaryTreeToLinkedList.java
package algorithms.treeandgraph; 

import java.util.*; 
import lib.TreeNode; 

public class FlattenBinaryTreeToLinkedList {
    /* Function: Let the tree nodes' right child point to its pre-order successor. 
     * Solution: Iterative pre-order. Keep a pointer on parent. 
     */ 
    
    public void flatten(TreeNode root) {
        if (root == null) {
            return; 
        }
        
        Stack<TreeNode> stack = new Stack<TreeNode>(); 
        stack.push(root); 
        TreeNode par = new TreeNode(0); 
        while (!stack.empty()) {
            TreeNode cur = stack.pop(); 
            par.right = cur; 
            
            if (cur.right != null) {
                stack.push(cur.right); 
            }
            if (cur.left != null) {
                stack.push(cur.left); 
            }
            
            par = cur; 
            cur.left = null; 
        }
    }
}

File: src/CPP/src/leetcode/FlattenBinaryTreeToLinkedList.cpp
/*
 * [Source] https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
 * [Difficulty]: Medium
 * [Tag]: Tree
 * [Tag]: Depth-first Search
 */

#include <iostream>

using namespace std;

// [Solution]: Preorder traverse the tree with stack, and then set right sub tree of previous linked node as the current node.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/treeandgraph/FlattenBinaryTreeToLinkedList.java
 */

/* Java solution
public class Solution {
	TreeNode last=null;
    public void flatten(TreeNode root) {
		if(root==null) return;
		if(last!=null){
			last.right=root;
			last=null;
		}
		TreeNode temp=root.right;
		if(root.left!=null){
			root.right=root.left;
			flatten(root.left);
			root.left=null;
		}
		flatten(temp);
		if(root.right==null) last=root;
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
